sdk/identity/azure-identity/src/managed_identity_credential.cpp

```diff
@@ -49,7 +51,38 @@ ManagedIdentityCredential::ManagedIdentityCredential(
     Azure::Core::Credentials::TokenCredentialOptions const& options)
     : TokenCredential("ManagedIdentityCredential")
 {
-  m_managedIdentitySource = CreateManagedIdentitySource(GetCredentialName(), clientId, {}, options);
+  m_managedIdentitySource
+      = CreateManagedIdentitySource(GetCredentialName(), clientId, {}, {}, options);
+}
+
+ManagedIdentityCredential::ManagedIdentityCredential(
+    Azure::Identity::ManagedIdentityCredentialOptions const& options)
+    : TokenCredential("ManagedIdentityCredential")
+{
+  ManagedIdentityIdType idType = options.IdentityType.GetManagedIdentityIdType();
+  if (idType == ManagedIdentityIdType::SystemAssigned)
+  {
+    m_managedIdentitySource = CreateManagedIdentitySource(GetCredentialName(), {}, {}, {}, options);
+  }
+  else if (idType == ManagedIdentityIdType::ClientId)
+  {
+    m_managedIdentitySource = CreateManagedIdentitySource(
+        GetCredentialName(), options.IdentityType.GetId(), {}, {}, options);
+  }
+  else if (idType == ManagedIdentityIdType::ObjectId)
+  {
+    m_managedIdentitySource = CreateManagedIdentitySource(
+        GetCredentialName(), {}, options.IdentityType.GetId(), {}, options);
+  }
+  else
```

Comment:
Would this make sense as a "switch" statement? That way you don't need the AZURE_ASSERT_MSG at line 79, and it allows the compiler to do a better job of optimizing the code (a smart compiler will figure out what you're doing, but why not make the compiler's job easier)?

Reply:
We'd still need the `AZURE_ASSERT_MSG` for the default swith case, right?

Reply:
Not really - since it's an enumeration, some compilers will actually complain if you have a `default` switch case (because it can never be reached). They also complain if you don't have all the enumeration variants covered, which future proofs adding in new variants.

Reply:
Interesting, that's cool.